File: MeatChat/MCPostListViewController.m
//
//  NMeatViewController.m
//  MeatChat
//
//  Created by Marcus Ramberg on 14.01.14.
//  Copyright (c) 2014 Nordaaker AS. 
//

#import "MCPostListViewController.h"
#import "MCPostCell.h"
#import <AVFoundation/AVFoundation.h>
#import "Reachability.h"
#import "MeatChat2-Swift.h"


@interface MCPostListViewController ()

@property (retain,nonatomic) NSMutableArray *items;
@property (retain,nonatomic) NSMutableDictionary *blocked;
@property (assign,nonatomic) BOOL atBottom;
@property (strong, nonatomic) NSMutableDictionary *seen;
@property (strong,nonatomic) NSString *userId;

@property (weak, nonatomic) IBOutlet NSLayoutConstraint *containerBottom;
@property (weak, nonatomic) IBOutlet UIView *containerView;
@property (weak, nonatomic) IBOutlet UILabel *activeCount;
@property (weak, nonatomic) IBOutlet UIButton *blockButton;
@property (assign, nonatomic) BOOL acceptedEula;


- (void)setupReachability;
- (void)setupSocket;
- (void)teardownSocket;
- (void)addPost: (NSDictionary*)data;
- (void)handleDisconnect;

- (void)keyboardWillHide:(NSNotification *)sender;
- (void)keyboardDidShow:(NSNotification *)sender;

@end

@implementation MCPostListViewController

#pragma mark - UITableview subclass methods

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.blocked=[[[NSUserDefaults standardUserDefaults] dictionaryForKey: @"meatspaceBlocks"] mutableCopy];
    if([self.blocked count]) {
        self.blockButton.hidden=NO;
    }
    self.acceptedEula=[[NSUserDefaults standardUserDefaults] boolForKey: @"acceptedEula"];
    if(!self.acceptedEula) {
        UIAlertController *alertController = [UIAlertController
                                              alertControllerWithTitle: @"Welcome to MeatChat"
                                              message: @"This is a client for the real time chat service chat.meatspac.es. Please behave nicely, objectionable content is not accepted. If someone else posts objectionable content, you can remove current and future messages from them, by using the 'block' button next to their message."
                                              preferredStyle:UIAlertControllerStyleAlert];
        
        
        UIAlertAction *cancelAction = [UIAlertAction
                                       actionWithTitle:NSLocalizedString(@"Cancel", @"Cancel action")
                                       style:UIAlertActionStyleCancel
                                       handler:^(UIAlertAction *action)
                                       {
                                           exit(0);
                                       }];
        
        UIAlertAction *okAction = [UIAlertAction
                                   actionWithTitle:NSLocalizedString(@"OK", @"OK action")
                                   style:UIAlertActionStyleDefault
                                   handler:^(UIAlertAction *action)
                                   {
                                       self.acceptedEula=YES;
                                       [[NSUserDefaults standardUserDefaults] setBool: self.acceptedEula forKey: @"acceptedEula"];
                                       
                                   }];
        [alertController addAction:cancelAction];
        [alertController addAction:okAction];
        
        [self presentViewController: alertController animated:YES completion:nil];
    }
    
    // The most pleasing inset 
    [self.tableView setContentInset:UIEdgeInsetsMake(42, 0, 0, 0)];
    
    self.items=[NSMutableArray array];
    self.seen=[NSMutableDictionary dictionary];
    [self setupReachability];
    
    // Keyboard handling
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc]
                                   initWithTarget:self
                                   action:@selector(dismissKeyboard)];
    
    [self.view addGestureRecognizer:tap];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillHide:) name:UIKeyboardWillHideNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardDidShow:) name:UIKeyboardDidShowNotification object:nil];
    
    // Tableview
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.estimatedRowHeight=75;
    self.atBottom=YES;
}


-(void)scrollToBottom
{
  NSIndexPath *indexPath=[NSIndexPath indexPathForItem:[self.items count]-1 inSection:0];
  [self.tableView scrollToRowAtIndexPath: indexPath atScrollPosition:UITableViewScrollPositionBottom animated:YES];
}

-(void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
{
  self.atBottom = NO;
  for (MCPostCell *cell in self.tableView.visibleCells) {
    [cell.videoPlayer pause];
  }
}

- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate
{
  [self endScroll: scrollView];
}

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView
{
  [self endScroll: scrollView];
}

- (void)resumePlay
{
  for (MCPostCell *cell in self.tableView.visibleCells) {
    [cell.videoPlayer play];
  }
}

- (void)endScroll: (UIScrollView*)scrollView
{
  [self resumePlay];
  CGFloat height = scrollView.frame.size.height;
  CGFloat contentYoffset = scrollView.contentOffset.y;
  CGFloat distanceFromBottom = scrollView.contentSize.height - contentYoffset;
  self.atBottom = (distanceFromBottom <= height);
}

- (void)scrollViewDidEndScrollingAnimation:(UIScrollView *)scrollView
{
  for (MCPostCell *cell in self.tableView.visibleCells) {
    [cell.videoPlayer play];
  }
  [self resumePlay];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
  if ([segue.identifier isEqualToString:@"postViewSegue"]) {
    // can't assign the view controller from an embed segue via the storyboard, so capture here
    _postViewController = (MCPostViewController*)segue.destinationViewController;
  }
}


#pragma mark - Socket handling

- (void)setupReachability;
{
  NSURL *server_url=[NSURL URLWithString: [[NSUserDefaults standardUserDefaults] objectForKey: @"server_url"]];
  Reachability* reach = [Reachability reachabilityWithHostname:[server_url host]];
  [[NSNotificationCenter defaultCenter] addObserver:self
                                           selector:@selector(reachabilityChanged:)
                                               name:kReachabilityChangedNotification
                                             object:nil];
  [reach startNotifier];
  [self setupSocket];
  
}

- (void)reachabilityChanged: (NSNotification*)notif
{
  Reachability *reach=notif.object;
  reach.isReachable ? [self setupSocket] : [self teardownSocket];
}

- (void)setupSocket
{
   __weak typeof(self) weakSelf = self;
  NSLog(@"%@",[[NSUserDefaults standardUserDefaults] objectForKey: @"server_url"]);
  [self.postViewController setPlaceholder: @"Connecting to meatspace"];
  [SIOSocket socketWithHost: [[NSUserDefaults standardUserDefaults] objectForKey: @"server_url"]
                   response: ^(SIOSocket *socket)
   {
     self.socket = socket;
     self.socket.onConnect = ^() {
       [weakSelf.socket emit: @"join" args: @[ @"mp4" ]];
       dispatch_async(dispatch_get_main_queue(), ^{
         weakSelf.postViewController.textfield.enabled = YES;
         [weakSelf.postViewController setRandomPlaceholder];
       });
     };
     self.socket.onDisconnect= ^() {
       dispatch_async(dispatch_get_main_queue(), ^{ [weakSelf handleDisconnect]; });
     };
     [self.socket on: @"message" callback:^(id data) {
       dispatch_async(dispatch_get_main_queue(), ^{ [weakSelf addPost: data]; });
     }];
     [self.socket on: @"messageack" callback:^(NSArray *data) {
       NSString *message=data[0];
       dispatch_async(dispatch_get_main_queue(), ^{
         if([[message class] isSubclassOfClass: [NSString class]]) {
           [weakSelf.postViewController setPlaceholder: message];
         };
         self.userId=[data[1] objectForKey: @"userId"];
       });
     }];
     [self.socket on: @"active" callback:^(NSArray *args) {
       dispatch_async(dispatch_get_main_queue(), ^{
         self.activeCount.text=[args[0] stringValue];
         self.activeCount.hidden=NO;
       });
     }];
     self.socket.onError = ^(NSDictionary *errorInfo) {
       NSLog(@"Oops: %@",errorInfo);
       dispatch_async(dispatch_get_main_queue(), ^{
         [weakSelf.postViewController setPlaceholder: [NSString stringWithFormat: @"An error occured: %@",errorInfo]];
       });
     };
     self.socket.onReconnect = ^(NSInteger numberOfAttempts) {
       NSLog(@"Reconnect %ld", (long)numberOfAttempts);
     };
     self.socket.onReconnectionAttempt =^(NSInteger numberOfAttempts) {
       NSLog(@"Attempt %ld", (long)numberOfAttempts);
       dispatch_async(dispatch_get_main_queue(), ^{
         [weakSelf.postViewController setPlaceholder: @"Reconnecting to meatspace."];
       });
     };
     self.socket.onReconnectionError=^(NSDictionary *errorInfo) {
       NSLog(@"Oops: %@",errorInfo);
       dispatch_async(dispatch_get_main_queue(), ^{
         [weakSelf.postViewController setPlaceholder: [NSString stringWithFormat: @"Could not connect: %@", errorInfo]];
       });
     };
   }];
}

- (void)teardownSocket
{
  [self.socket close];
  self.socket=nil;
  [self.postViewController setPlaceholder: @"Get the internet, bae."];
}


-(void)handleDisconnect
{
   [self.postViewController.textfield resignFirstResponder];
   self.postViewController.textfield.enabled=NO;
   [self.postViewController setPlaceholder: @"Disconnected, please hold"];
   self.activeCount.hidden=YES;
  
}

- (void)flushItems
{
  for (MCPost *item in self.items) {
    [item cleanup];
  }
  [self.items removeAllObjects];
  [self.tableView reloadData];
}

- (void)addPost: (NSArray*)args
{
  NSDictionary *data=args[0];
  if([self.blocked objectForKey: [data objectForKey: @"fingerprint"]]) { return; }
  
  // Flush old posts
  [self.tableView beginUpdates];
  for( int i = (int)[self.items count] - 1; i >= 0; --i) {
    MCPost *post=[self.items objectAtIndex: i];
    if ([post isObsolete]) {
      [post cleanup];
      [self.items removeObjectAtIndex: i];
      [self.tableView deleteRowsAtIndexPaths:@[[NSIndexPath indexPathForItem:i inSection:0]]  withRowAnimation:UITableViewRowAnimationAutomatic];
    }
  }
  [self.tableView endUpdates];
  
  NSString *key=[data objectForKey: @"key"];
  if (key) {
    if([self.seen objectForKey: key] ) { return; }
    [self.seen setObject: @"1" forKey: key];
  };
  MCPost *post=[[MCPost alloc] initWithDict: data];
  [self.items addObject: post];
  NSIndexPath *newRow=[NSIndexPath indexPathForItem:[self.items count]-1 inSection:0];
  [CATransaction begin];
  [CATransaction setCompletionBlock:^{
    if (self.atBottom) {
      [self scrollToBottom];
    }
  }];
  [self.tableView beginUpdates];
  [self.tableView insertRowsAtIndexPaths:@[newRow] withRowAnimation: UITableViewRowAnimationFade];
  [self.tableView endUpdates];
  [CATransaction commit];
}



#pragma mark - Keyboard handling

-(void)dismissKeyboard {
  [self.postViewController closePostWithPosted: NO];
}

- (void)keyboardDidShow:(NSNotification *)sender {
  CGRect frame = [sender.userInfo[UIKeyboardFrameEndUserInfoKey] CGRectValue];
  self.containerBottom.constant = frame.size.height;
  [self.containerView setNeedsUpdateConstraints];
  [UIView animateWithDuration:0.25f animations:^{
    [self.containerView layoutIfNeeded];
    self.containerHeight.constant = 75.0;
    self.postViewController.characterCount.hidden=NO;
  } completion:^(BOOL finished) {
    if([self.items count]) {
      [self scrollToBottom];
      self.atBottom=YES;
    }
  }];
}

- (void)keyboardWillHide:(NSNotification *)sender
{
  self.containerBottom.constant = 0;
  [self.containerView setNeedsUpdateConstraints];
  [UIView animateWithDuration:0.25f animations:^{
    [self.containerView layoutIfNeeded];
    self.containerHeight.constant = 35.0;
  } completion:^(BOOL finished) {
    if([self.items count]) {
      [self scrollToBottom];
      [self.tableView reloadRowsAtIndexPaths: @[[NSIndexPath indexPathForItem:[self.items count]-1 inSection:0]]
                            withRowAnimation: UITableViewRowAnimationNone];
      self.atBottom=YES;
    }
  }];
}


- (IBAction)unblockClicked:(id)sender {
  [self.blocked removeAllObjects];
  [[NSUserDefaults standardUserDefaults] setObject: self.blocked forKey:@"meatspaceBlocks"];
  self.blockButton.hidden=YES;
}

- (IBAction)blockClicked:(UIButton*)sender {
  MCPost *blockPost=[self.items objectAtIndex: sender.tag];
  [self.blocked setObject: @"1" forKey: blockPost.fingerprint];
  self.blockButton.hidden=NO;
  [[NSUserDefaults standardUserDefaults] setObject: self.blocked forKey:@"meatspaceBlocks"];
  for( int i = (int)[self.items count]-1; i >=0; --i) {
    MCPost *post=[self.items objectAtIndex: i];
    if([post.fingerprint isEqualToString: blockPost.fingerprint] ) {
      [self.items removeObject: post];
    }
    [self.tableView reloadData];
  }
  __weak __typeof(self)weakSelf = self;
    UIAlertController *alertController = [UIAlertController
                                          alertControllerWithTitle: @"Report abuse?"
                                          message: @"You have just chosen to block another meatspacer. Would you like to report this user for objectionable content?"
                                          preferredStyle:UIAlertControllerStyleAlert];
    
    
    UIAlertAction *cancelAction = [UIAlertAction
                                   actionWithTitle:NSLocalizedString(@"No", @"No")
                                   style:UIAlertActionStyleCancel
                                   handler:^(UIAlertAction *action)
                                   {
                                   }];
    
    UIAlertAction *okAction = [UIAlertAction
                               actionWithTitle:NSLocalizedString(@"Yes", @"Yes")
                               style:UIAlertActionStyleDefault
                               handler:^(UIAlertAction *action)
                               {
                                 MFMailComposeViewController *mf=[[MFMailComposeViewController alloc] init];
                                 [mf setToRecipients: @[@"report@meatspac.es"]];
                                 mf.mailComposeDelegate=weakSelf;
                                 [mf setSubject: [NSString stringWithFormat: @"Abuse from user fingerprint %@",blockPost.fingerprint]];
                                [weakSelf presentViewController: mf animated: YES completion:^{
                                }];
                                   
                               }];
    [alertController addAction:cancelAction];
    [alertController addAction:okAction];
    
    [self presentViewController: alertController animated:YES completion:nil];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    return [self.items count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
  static NSString *CellIdentifier = @"MeatCell";
  MCPostCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];
  
  // Configure the cell...
  MCPost *post=[self.items objectAtIndex:indexPath.row];
  cell.textView.attributedText=post.attributedString;
  cell.timeLabel.text=[post relativeTime];
  if([self.userId isEqualToString: post.fingerprint]) {
    self.blockButton.hidden=YES;
  }
  else {
    cell.blockButton.hidden=NO;
    cell.blockButton.tag=indexPath.row;
  }
  AVPlayerItem *item=[AVPlayerItem playerItemWithURL: post.videoUrl];

  [cell.videoPlayer replaceCurrentItemWithPlayerItem: item];
  
  if(([self.items count]-1)==[self.tableView.visibleCells count]) {
    [cell.videoPlayer play];
  }
  [[NSNotificationCenter defaultCenter] addObserver: cell
                                           selector:@selector(playerItemDidReachEnd:)
                                               name:AVPlayerItemDidPlayToEndTimeNotification
                                             object:item];
  return cell;
}

- (void)tableView:(UITableView *)tableView didEndDisplayingCell:(UITableViewCell *)cell forRowAtIndexPath:(NSIndexPath *)indexPath
{
  [[(MCPostCell*)cell videoPlayer] pause];
}

- (void)tableView:(UITableView *)tableView didDeselectRowAtIndexPath:(NSIndexPath *)indexPath
{
  MCPostCell *cell = (MCPostCell*)[tableView cellForRowAtIndexPath:indexPath];
  cell.blockButton.hidden=YES;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
  MCPostCell *cell = (MCPostCell*)[tableView cellForRowAtIndexPath:indexPath];
  cell.blockButton.hidden=NO;
}


#pragma mark - AVPlayer delegate




#pragma mark - UITextView delegate for cells

- (BOOL)textView:(UITextView *)textView shouldInteractWithURL:(NSURL *)URL inRange:(NSRange)characterRange
{
  return YES;
}

- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error
{
  [self dismissViewControllerAnimated: YES completion:^{
    
  }];
}

@end

File: MeatChat/MCPostListViewController.swift
//
//  MCPostListViewController.swift
//  MeatChat2
//
//  Created by Marcus Ramberg on 22/03/16.
//  Copyright © 2016 Nordaaker AS. All rights reserved.
//

import UIKit
import MessageUI.MFMailComposeViewController
import SocketIO
import ReachabilitySwift

class MCPostListViewController : UIViewController, UITableViewDataSource, UITableViewDelegate, UITextViewDelegate, AVAssetResourceLoaderDelegate,MFMailComposeViewControllerDelegate {
    
   
    var items:NSMutableArray=[]
    var blocked:[String: AnyObject]=[:]
    var atBottom:Bool=false
    var seen:[String: AnyObject]=[:]
    var userId:String=""
    
    @IBOutlet weak var containerHeight:NSLayoutConstraint!
    @IBOutlet weak var tableView:UITableView!
    @IBOutlet weak var containerBottom:NSLayoutConstraint!
    @IBOutlet weak var containerView:UIView!
    @IBOutlet weak var activeCount:UILabel!
    @IBOutlet weak var blockButton:UIButton!

    var  socket:SocketIOClient?
    var  postViewController:MCPostViewController?
    var acceptedEula:Bool!
    
    override func viewDidLoad() {
        if let blockList=UserDefaults.standard.dictionary(forKey: "meatspaceBlocks") {
            blocked=blockList as [String : AnyObject]
        }
        if((blocked.count) > 0) {
            blockButton.isHidden=false
        }
        acceptedEula=UserDefaults.standard.bool(forKey: "acceptedEula");
        if(!self.acceptedEula) {
            let eulaAlert = UIAlertController(title: "Welcome to MeatChat", message: "This is a client for the real time chat service chat.meatspac.es. Please behave nicely, objectionable content is not accepted. If someone else posts objectionable content, you can remove current and future messages from them, by using the 'block' button next to their message.", preferredStyle: UIAlertControllerStyle.alert)
            let cancelAction = UIAlertAction(title: "Cancel", style: UIAlertActionStyle.cancel, handler: { (UIAlertAction) -> Void in
                exit(0);
            })
            eulaAlert.addAction(cancelAction)
            let OKAction = UIAlertAction(title: "OK", style: UIAlertActionStyle.default, handler: { (UIAlertAction) -> Void in
                self.acceptedEula=true;
                UserDefaults.standard.set(self.acceptedEula, forKey: "acceptedEula")
            })
            eulaAlert.addAction(OKAction)
            self.present(eulaAlert, animated: true, completion: nil)
        }
        
        tableView.contentInset = UIEdgeInsetsMake(42, 0, 0, 0);
        
        self.setupReachability()
        
        let tap = UITapGestureRecognizer(target: self, action: #selector(UIInputViewController.dismissKeyboard))
        self.view.addGestureRecognizer(tap)
        NotificationCenter.default.addObserver(self, selector: #selector(MCPostListViewController.keyboardWillHide(_:)), name: NSNotification.Name.UIKeyboardWillHide, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(MCPostListViewController.keyboardDidShow(_:)), name: NSNotification.Name.UIKeyboardDidShow, object: nil)
        
        self.tableView.rowHeight = UITableViewAutomaticDimension;
        self.tableView.estimatedRowHeight=75;

    }
    
    func scrollToBottom() {
        let indexPath=IndexPath(item: (self.items.count-1), section: 0)
        self.tableView.scrollToRow(at: indexPath, at: UITableViewScrollPosition.bottom, animated: true)
    }
    
    func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        self.atBottom = false;
        for cell in self.tableView.visibleCells {
            if let postCell = cell as? MCPostCell {
                postCell.videoPlayer!.pause()
            }
        }

    }
   
    func resumePlay() {
        for cell in self.tableView.visibleCells {
            if let postCell = cell as? MCPostCell {
                postCell.videoPlayer!.play()
            }
        }
        
    }
    
    func endScroll(_ scrollView:UIScrollView) {
        self.resumePlay();
        let height = scrollView.frame.size.height;
        let contentYoffset = scrollView.contentOffset.y;
        let distanceFromBottom = scrollView.contentSize.height - contentYoffset;
        self.atBottom = (distanceFromBottom <= height);

    }
    
    func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        self.endScroll(scrollView)
    }
    
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        self.endScroll(scrollView)
    }
    
    func scrollViewDidEndScrollingAnimation(_ scrollView: UIScrollView) {
        self.resumePlay()
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let ctrl=segue.destination as? MCPostViewController {
            self.postViewController=ctrl
        }
    }
    
    func setupReachability() {
        let server_url=URL(string: UserDefaults.standard.object(forKey: "server_url") as! String);
        let reach = Reachability(hostname: server_url!.host!)!
        NotificationCenter.default.addObserver(self, selector: #selector(MCPostListViewController.reachabilityChanged(_:)), name: ReachabilityChangedNotification, object: nil)
        do {
            try reach.startNotifier()
        } catch {
            print("could not start reachability notifier")
        }
        // self.setupSocket();
    }
    
    func reachabilityChanged(_ notif:Notification) {
        let reach=notif.object as! Reachability;
        reach.isReachable ? self.setupSocket() : self.teardownSocket();
    }
    

        
    func teardownSocket() {
        
        self.socket!.disconnect()
        self.socket=nil;
        self.postViewController!.setPlaceHolder("Get the internet, bae.")
    }
    
    func handleDisconnect() {
        self.postViewController!.textfield.resignFirstResponder()
        self.postViewController!.textfield.isEnabled=false
        self.postViewController!.setPlaceHolder("Disconnected, please hold")
        self.activeCount.isHidden = true 

    }
        
    func flushItems() {
        for item in self.items {
            if let postItem = item as? MCPost {
                postItem.cleanup()
            }
        }
        self.items.removeAllObjects()
        self.tableView.reloadData()
    }
        
    func addPost(_ args:NSArray) {
        if let fingerprint=(args[0] as! NSDictionary)["fingerprint"] as? String {
            if(self.blocked.keys.contains(fingerprint)) { return; }
        
            self.tableView.beginUpdates()
            if self.items.count > 0 {
                for  i in 0 ... (self.items.count - 1)  {
                    if let post=self.items.object(at: i) as? MCPost {
                        if post.isObsolete() {
                            post.cleanup()
                            self.items.removeObject(at: i)
                            let indexPath=IndexPath(item: i, section: 0)
                            self.tableView.deleteRows(at: [indexPath], with:UITableViewRowAnimation.automatic)
                        }
                    }
                }
            }
            self.tableView.endUpdates()
        
            if let key = (args[0] as! NSDictionary)["key"] as? String {
                if((self.seen[key] ) != nil) { return; }
                self.seen[key]="1" as AnyObject?
            }
        }
        let post=MCPost(dict: args[0] as! NSDictionary)
        self.items.add(post)
        let newRow=IndexPath(item: self.items.count-1, section: 0)

        CATransaction.begin()
        CATransaction.setCompletionBlock({
            if (self.atBottom != true) {
                self.scrollToBottom()
            }
        })
        self.tableView.beginUpdates()
        self.tableView.insertRows(at: [newRow], with: UITableViewRowAnimation.fade)
        self.tableView.endUpdates()
        CATransaction.commit()
    }

    func dismissKeyboard() {
        self.postViewController!.closePostWithPosted(false)
    }

    func keyboardDidShow(_ sender:Notification) {

        let frame = (sender.userInfo![UIKeyboardFrameEndUserInfoKey]! as AnyObject).cgRectValue
        self.containerBottom.constant = (frame?.size.height)!
        self.containerView.setNeedsUpdateConstraints()
        UIView.animate(withDuration: 0.25, animations: {
            self.containerView.layoutIfNeeded()
            self.containerHeight.constant = 75.0
            self.postViewController!.characterCount.isHidden=false
        }, completion: { (finished) in
            if self.items.count > 0 {
                self.scrollToBottom()
                self.atBottom=true
            }
        }) 
        
    }
    func keyboardWillHide(_ sender:Notification) {
        self.containerBottom.constant = 0;
        self.containerView.setNeedsUpdateConstraints()
        UIView.animate(withDuration: 0.25, animations: {
            self.containerView.layoutIfNeeded()
            self.containerHeight.constant = 35.0
        }, completion: { (finished) in
            if self.items.count > 0 {
                self.scrollToBottom()
                let indexPath=IndexPath(item: self.items.count-1, section: 0)
                self.tableView.reloadRows( at: [indexPath], with: UITableViewRowAnimation.none )
                self.atBottom=true
            }
        }) 
    }
     
    @IBAction func unblockClicked(_ sender:AnyObject) {
        self.blocked=[:]
        UserDefaults.standard.set(self.blocked, forKey:"meatspaceBlocks")
        self.blockButton.isHidden=true

    }
    
    @IBAction func blockClicked(_ sender:AnyObject) {
        if let blockPost=self.items.object(at: sender.tag) as? MCPost {
            self.blocked[blockPost.fingerprint]="1" as AnyObject
            self.blockButton.isHidden=false
            UserDefaults.standard.set(self.blocked, forKey: "meatspaceBlocks")
            for i in 0 ... self.items.count-1{
                if let post = self.items.object(at: i) as? MCPost {
                    if post.fingerprint == blockPost.fingerprint {
                        self.items.remove(post)
                        self.tableView.reloadData()
                    }
                }
            }
            weak var weakSelf=self
            let alertController = UIAlertController(title: "Report abuse?", message: "You have just chosen to block another meatspacer. Would you like to report this user for objectionable content?", preferredStyle: UIAlertControllerStyle.alert)
            let cancelAction=UIAlertAction(title: "No", style: UIAlertActionStyle.cancel) { (action) in
                
            }
            
            let okAction=UIAlertAction(title: "Yes", style: UIAlertActionStyle.default) { (action) in
                let mf=MFMailComposeViewController()
                mf.setToRecipients(["report@meatspac.es"]);
                mf.mailComposeDelegate=weakSelf;
                mf.setSubject(String(format: "Abuse from user fingerprint %@",blockPost.fingerprint));
                weakSelf!.present(mf, animated: true, completion: {})
                
            }
            alertController.addAction(cancelAction)
            alertController.addAction(okAction)
            
            self.present(alertController, animated:true, completion:nil);
        }

        
        

    }

    func setupSocket() {
        weak var weakSelf=self
        self.postViewController!.setPlaceHolder("Connecting to meatspace")

        
        print("setupSocket")
        self.socket = SocketIOClient(socketURL: URL(string: UserDefaults.standard.object(forKey: "server_url") as! String)!)

        self.socket!.on(clientEvent: .connect) { data, ack in
            print("join")
            weakSelf?.socket!.emit("join", ["mp4"])
            DispatchQueue.main.async(execute: {
                weakSelf?.postViewController!.textfield.isEnabled=true
                weakSelf?.postViewController!.setRandomPlaceholder()
            })
            ack.with("connected")
        }
        self.socket!.on(clientEvent: .disconnect) { data, ack in
            DispatchQueue.main.async(execute: {
                weakSelf?.handleDisconnect()
            })
        }
        self.socket!.on("chat") { data, ack in
            let arr = data as NSArray
            print(arr.count)
            DispatchQueue.main.async(execute: {
                print(arr[0] as! NSDictionary)
//                print(data[1] as! NSDictionary)
            })
            
//            DispatchQueue.main.async(execute: {
//                if let arr = data as NSArray? {
//                    weakSelf!.addPost(arr)
//                }
//            })
        }
        self.socket!.on("message") { data, ack in
            print(data)
            DispatchQueue.main.async(execute: {
                let arr = data as NSArray
                weakSelf!.addPost(arr)
            })
        }
        self.socket!.on("messageack") { data, ack in
            if let message = (data as NSArray)[0] as? String {
                DispatchQueue.main.async(execute: {
                    weakSelf?.postViewController!.setPlaceHolder(message)
                    if let uid=((data as NSArray)[1] as! NSDictionary)["userId"] as? String {
                        self.userId=uid
                    }
                })
            }
        }
        self.socket!.on("userid") { data, ack in
            print(data)
            DispatchQueue.main.async(execute: {
                if let uid=((data as NSArray)[1] as! NSDictionary)["userId"] as? String {
                    self.userId=uid
                }
            })
        }
        self.socket!.on("active") { args, ack in
            DispatchQueue.main.async(execute: {
                self.activeCount.text=((args as NSArray)[0] as AnyObject).stringValue
                self.activeCount.isHidden=false
            })
            
        }
        self.socket!.on(clientEvent: .error) { (errorInfo, ack) in
            print(errorInfo)
            DispatchQueue.main.async(execute: {
                weakSelf?.postViewController!.setPlaceHolder(String(format: "An error occured: %@", errorInfo))
            })
        }
        self.socket!.on(clientEvent: .reconnect) { numberOfAttempts, ack in
            print(String(format: "Reconnect %ld",numberOfAttempts))
        }
        self.socket!.on(clientEvent: .reconnectAttempt) { numberOfAttempts, ack in
            print(String(format: "Attempt %ld",numberOfAttempts))
            DispatchQueue.main.async(execute: {
                weakSelf?.postViewController!.setPlaceHolder("Reconnecting to meatspace.")
            })
        }
//        self.socket!.onReconnectionError={ (errorInfo) in
//            print(errorInfo)
//            DispatchQueue.main.async(execute: {
//                weakSelf!.postViewController!.setPlaceHolder(String(format: "Could not connect: %@", errorInfo!))
//            })
//        }
        
        self.socket!.connect()
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.items.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if let cell=tableView.dequeueReusableCell(withIdentifier: "MeatCell", for: indexPath) as? MCPostCell {
            if let post=self.items.object(at: indexPath.row) as? MCPost {
                cell.textView.attributedText=post.attributedString
                cell.timeLabel.text=post.relativeTime()
                if self.userId == post.fingerprint {
                    self.blockButton.isHidden=true
                } else {
                    cell.blockButton.isHidden=false
                    cell.blockButton.tag=indexPath.row
                }
                let item=AVPlayerItem(url: post.videoUrl as URL)
                cell.videoPlayer?.replaceCurrentItem(with: item)
                if self.items.count-1 == self.tableView.visibleCells.count {
                    cell.videoPlayer?.play()
                }
                NotificationCenter.default.addObserver(cell, selector: #selector(cell.playerItemDidReachEnd(_:)), name: NSNotification.Name.AVPlayerItemDidPlayToEndTime, object: item)
            }
        return cell
        }
        return UITableViewCell()
    }
    
    func tableView(_ tableView: UITableView, didEndDisplaying cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        if let postCell=cell as? MCPostCell {
            postCell.videoPlayer!.pause()
        }
    }
    
    func tableView(_ tableView: UITableView, didDeselectRowAt indexPath: IndexPath) {
        if let cell=tableView.cellForRow(at: indexPath) as? MCPostCell {
            cell.blockButton.isHidden=true
        }
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if let cell=tableView.cellForRow(at: indexPath) as? MCPostCell {
            cell.blockButton.isHidden=false
        }
    }
    
    func textView(_ textView: UITextView, shouldInteractWith URL: URL, in characterRange: NSRange) -> Bool {
        return true
    }
    
    func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        self.dismiss(animated: true) {}
    }

}

